src/pymor/algorithms/rand_la.py

```diff
@@ -2,21 +2,266 @@
+
+    def _estimate_error(self, basis_size, num_testvecs, p_fail):
+        c = np.sqrt(2 * self._lambda_min())
+        c *= erfinv((p_fail / min(self.A.source.dim, self.A.range.dim)) ** (1 / num_testvecs))
+
+        if len(self._samplevecs) < num_testvecs:
+            self._draw_test_vector(num_testvecs - len(self._samplevecs))
+
+        W, Q = self._samplevecs[:num_testvecs].copy(), self._find_range(basis_size)
+        W -= Q.lincomb(Q.inner(W, self.range_product).T)
+        return c * np.max(W.norm(self.range_product))
+
+    def estimate_error(self, basis_size, num_testvecs=20, p_fail=1e-14):
+        r"""Randomized a posteriori error estimator for a given basis size.
+
+        This implements the a posteriori error estimator from :cite:`BS18` (Definition 3.1).
+
+        The error is given by
+
+        .. math::
+            \epsilon_{\mathrm{est}}=c_{\mathrm{est}}\cdot\max_{\omega\in\Omega}
+            \lVert (I-QQ^T)A\omega\rVert_{T\rightarrow S}
+
+        with
+
+        .. math::
+            c_{\mathrm{est}}
+            =\frac{1}{\sqrt{2\lambda_{\mathrm{min}}}\operatorname{erf}^{-1}
+            \left(\left(\frac{\texttt{p_fail}}{n}\right)^{1/\texttt{num_testvecs}}\right)},
+
+        where :math:`\Omega` is a set of `num_testvecs` random vectors, :math:`S` denotes the inner
+        product of the range, :math:`T` denotes inner product of the source of :math:`A` and
+        :math:`n=\min\{\dim(S),\,\dim(T)\}`.
+
+        Parameters
+        ----------
+        basis_size
+            The size of the basis.
+        num_testvecs
+            Number of test vectors for error estimation.
+        p_fail
+            Maximum failure probabilty of the error estimate.
+
+        Returns
+        -------
+        err
+            The approximate error of the basis.
+        """
+        assert isinstance(basis_size, int) and basis_size > 0
+        if basis_size > min(self.A.source.dim, self.A.range.dim):
+            self.logger.warning('Requested basis is larger than the rank of the operator!')
+            self.logger.info('Proceeding with maximum operator rank.')
+            basis_size = min(self.A.source.dim, self.A.range.dim)
+        assert 0 < num_testvecs and isinstance(num_testvecs, int)
+        assert 0 < p_fail < 1
+
+        err = self._estimate_error(basis_size, num_testvecs, p_fail)
+        self.logger.info(f'Estimated error (basis dimension {basis_size}): {err:.5e}.')
+        return err
+
+    def _extend_basis(self, k):
+        W = self.A.source.random(k, distribution='normal')
+        if self.complex:
+            W += 1j * self.A.source.random(k, distribution='normal')
+
+        offset = len(self._Q[0])
+        self._Q[0].append(self.A.apply(W))
+        gram_schmidt(self._Q[0], self.range_product, offset=offset, copy=False)
+
+        for i in range(self.subspace_iterations):
+            i = 2*i + 1
+
+            k = len(self._Q[i-1]) - offset  # check if GS removed vectors
```

Comment:
I think we need to check here for `k == 0`. If that happens, then `self._Q[i-1][-k:]` would select all vectors instead of none. Not sure if `k == 0` should raise an error. Probably?

Reply:
I actually can't remember now if I added this out of necessity or caution. I cannot construct an example where `k==0` so I wonder whether it would even be ok to completely remove this check. What do you think?

Reply:
Aha, I found the origin https://github.com/pymor/pymor/pull/1753#discussion_r982529559.